### PR TITLE
GEODE-4059: Changing protobuf handshake to not need communication mod…

### PR DESCRIPTION
**Pull request by WireBaron**
…e bytes

- Moved handshake proto messages into top level handshake.proto
- Changed protobuf communication mode to match leading byte of handshake
- Updated handshake handler and tests to handle changes
- Modified locator to accept protobuf communication mode in lieu of gossip

Signed-off-by: Galen O'Sullivan <gosullivan@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by WireBaron** [comment]
@galen-pivotal @PivotalSarge @upthewaterspout @bschuchardt

File: geode-core/src/main/java/org/apache/geode/distributed/internal/tcpserver/TcpServer.java [hunk 1]
```diff
@@ -379,89 +379,13 @@ private void processRequest(final Socket socket) {
...
-          handler.endResponse(request, startTime);
+        // read the first byte & check for an improperly configured client pool trying
+        // to contact a cache server
+        int firstByte = input.readUnsignedByte();
+        if (firstByte != CommunicationMode.ReservedForGossip.getModeNumber()) {
+          handleNonGossipConnection(socket, input, firstByte);
         } else {
-          // Close the socket. We can not accept requests from a newer version
-          rejectUnknownProtocolConnection(socket, gossipVersion);
+          processOneConnection(socket, startTime, input);
```
Comment (PivotalSarge) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/tcpserver/TcpServer.java]: Does the first byte need to be passed here, too?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/CommunicationMode.java [hunk 2]
```diff
@@ -22,6 +22,19 @@
...
+   */
+  ReservedForGossip((byte) 0, "Locator gossip version"),
+  /**
+   * For the new client-server protocol.
+   *
+   * Protobuf handshake messages are specially constructed so that this value will match the first
+   * byte sent, allowing clients to start protobuf connections with a handshake instead of
+   * communication mode bytes.
+   */
+  ProtobufClientServerProtocol((byte) 10, "Protobuf client"),
```
Comment (PivotalSarge) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/CommunicationMode.java]: Will this require coordination with the C# and Go drivers?

**Comment by WireBaron** [comment]
Fixed a few issues related to failing tests, rebased to latest develop and pushed the latest version over the previous change.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/tcpserver/TcpServer.java (hunk 1 above)
Comment (WireBaron) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/tcpserver/TcpServer.java]: No, in this case the first byte was the high order byte of the gossip version, and thus guaranteed to be 0.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/CommunicationMode.java (hunk 2 above)
Comment (WireBaron) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/CommunicationMode.java]: Kind of, they'll need to back out any magic byte they're currently sending us and just start each connection with a Handshake message.

**Pull request closed without merging**